### PR TITLE
T2-VOQ-Chassis: VS support

### PR DESCRIPTION
**Pull request by deepak-singhal0408**
#### Why I did it
Changes to ensure sonic_vs.img has everything required for it to be emulated as T2-VOQ-Chassis: Supervisor/Linecard..
With this change, 
1. The image could be used to emulate Supervisor or Linecard
2. the image could be used to emulate different Linecard HWSKus.

##### Work item tracking
- Microsoft ADO **(number only)**:
27402561
#### How I did it
Following Changes are made as part of this PR:
1. On VS image, Database containers to be recreated again upon reboot. This will help change single asic VS image to multi-asic(docker_image_ctl.j2).
2. Copy all Sup and Linecard HWSKUs directories under kvm_platform directory. Create lanemap.ini, coreportindexmap.ini, fabriclanemap.ini under each hwsku/Asic directory. Copy asic.conf file from each platform directories to their child HW_SKU directories under kvm platform directory (This will help emulate different HW-Sku's with their respective num_asic asics):  (sonic-device-data/Makefile). 
3. New sonic-platform package for VS platforms. This is needed
 3.1:  To ensure, that database containers bring up goes through. The database containers on linecards fetch the supervisor slot_num, current_slot num etc.
PS: This package will be available on pizza box vs platforms as well. However, it will be noop there, as the package expects a metadata file, which will only be available on chassis VS platforms.
4. Changes to generate unique mac-address on VS platforms. This is done by using the device_hostname string (which will always be unique). Change to provide unique MAC address per asic on multi-asic VOQ VS platforms(sonic-cfggen, device_info.py, minigraph.py)
5. topology.service file to be dependent on sonic.target so that as part of config load_minigraph/config reload this service gets invoked.
6. topology.sh changes to move ports to their respective namespace (only applicable on multi-asic platforms)

#### How to verify it
1. Verified that sonic-vs.img.gz gets built succesfully and single asic VS DUT comes up fine(deployed vms-kvm-t0 topology) and  could see that all containers come up fine. The show commands work as expected.
2. Verified bringup on VOQ chassis with different flavor of linecard emulation.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

- [ ] 
- [ ]

### CODE REVIEW

File: files/build_templates/docker_image_ctl.j2 [hunk 1]
```diff
@@ -404,28 +404,33 @@ start() {
 
     DOCKERCHECK=`docker inspect --type container ${DOCKERNAME} 2>/dev/null`
     if [ "$?" -eq "0" ]; then
-        {%- if docker_container_name == "database" %}
-        DOCKERMOUNT=""
+        {%- if docker_container_name == "database" and sonic_asic_platform == "vs" %}
```
Comment (arlakshm) [inline files/build_templates/docker_image_ctl.j2:407]: fix indent? Why do we need to make this change? This means every time a vs switch is rebooted, the database switch will be recreated?

File: files/image_config/topology/topology.sh [hunk 2]
```diff
@@ -22,12 +22,99 @@ get_hwsku() {
...
+                    num="${prev#eth}"
+                    # Increment the numeric part
+                    ((num++))
+                    # Construct the new interface name
+                    cur="eth$num"
+                    echo "Renaming $intf to $cur"
+                    ip link set dev $intf down
+                    ip link set dev $intf name $cur
+                    ip link set dev $cur up
+               fi
```
Comment (arlakshm) [inline files/image_config/topology/topology.sh:101]: indention is not correct?

File: files/image_config/topology/topology.sh [hunk 3]
```diff
@@ -38,8 +125,10 @@ stop() {
     TOPOLOGY_SCRIPT="topology.sh"
     PLATFORM=`sonic-cfggen -H -v DEVICE_METADATA.localhost.platform`
     HWSKU=`get_hwsku`
-    if [[ $HWSKU != "" ]]; then
+    if [[ $HWSKU == "msft_"* ]]; then
```
Comment (arlakshm) [inline files/image_config/topology/topology.sh:128]: why do need to check of specific skus here?

File: files/image_config/topology/topology.sh [hunk 4]
```diff
@@ -22,12 +22,99 @@ get_hwsku() {
...
+    BIND="$3"
+
+    # Check if the directory exists
+    if [[ ! -d "/usr/share/sonic/device/$PLATFORM/$HWSKU" ]]; then
+        echo "Directory /usr/share/sonic/device/$PLATFORM/$HWSKU does not exist"
+        exit 1
+    fi
+
+    # Read NUM_ASIC from asic.conf file
+    asic_conf="/usr/share/sonic/device/$PLATFORM/asic.conf"
```
Comment (arlakshm) [inline files/image_config/topology/topology.sh:37]: we cannot assume all the LCs in T2 chassis is multi-asics. We need to handle single asic LC as well. For single asic LC asic.conf will not be present.

File: src/sonic-py-common/sonic_py_common/device_info.py [hunk 5]
```diff
@@ -687,16 +693,33 @@ def run_command_pipe(cmd0, cmd1, cmd2):
         err = out
     return (out, err)
 
+def _modify_mac_for_asic(mac, namespace=None):
+    if namespace is None:
+        return mac
+    if namespace in ["asic0", "asic1", "asic2"]:
```
Comment (arlakshm) [inline src/sonic-py-common/sonic_py_common/device_info.py:699]: can we get the namespace list iso hardcoding the asic names

**Review by arlakshm — COMMENTED** [review]
as comments.

File: files/build_templates/docker_image_ctl.j2 (hunk 1 above)
Comment (deepak-singhal0408) [reply files/build_templates/docker_image_ctl.j2:407]: Indentation seems fine to me? Should it be tab?

yes, this change means database container will be recreated everytime after reboot.. This is to support converting single asic image to multi-asic(based on asic.conf)..That requires database containers to be recreated again.

File: files/image_config/topology/topology.sh (hunk 4 above)
Comment (deepak-singhal0408) [reply files/image_config/topology/topology.sh:37]: this script gets executed from topology.service.. which is only instantiated on multi-asic devices..

File: files/image_config/topology/topology.sh (hunk 3 above)
Comment (deepak-singhal0408) [reply files/image_config/topology/topology.sh:128]: This is to keep our existing support of multi-asic pizza box DUTs.. for that, we had introduced msft_* hw_skus..

File: src/sonic-py-common/sonic_py_common/device_info.py (hunk 5 above)
Comment (deepak-singhal0408) [reply src/sonic-py-common/sonic_py_common/device_info.py]: Fixed..

File: files/image_config/topology/topology.sh (hunk 4 above)
Comment (arlakshm) [reply files/image_config/topology/topology.sh:37]: For single asic voq linecard we do not have to create the netdev for Cpu0?

**Comment by saiarcot895** [comment]
Looks like there's a new loganalyzer error message on t0:

```
Apr  8 07:23:49.040082 vlab-01 ERR sfputil: Failed to instantiate Chassis due to FileNotFoundError('Metadata file /etc/sonic/vs_chassis_metadata.json not found')
```

Is this expected? Does it need to be added to the ignore list?

**Comment by deepak-singhal0408** [comment]
> Looks like there's a new loganalyzer error message on t0:
> 
> ```
> Apr  8 07:23:49.040082 vlab-01 ERR sfputil: Failed to instantiate Chassis due to FileNotFoundError('Metadata file /etc/sonic/vs_chassis_metadata.json not found')
> ```
> 
> Is this expected? Does it need to be added to the ignore list?

Thanks @saiarcot895 .. Added above to loganalyzer_ignore file as discussed. PR https://github.com/sonic-net/sonic-mgmt/pull/12345

**Comment by deepak-singhal0408** [comment]
> Commenter does not have sufficient privileges for PR 18512 in repo sonic-net/sonic-buildimage

Hi @rlhui @yxieca , it seems I dont have permission to re run the pipeline? Could you please help check and let me know if this is expected or am I missing anything here?

**Comment by saiarcot895** [comment]
@deepak-singhal0408 use `/azpw` instead of `/azp`.

**Comment by deepak-singhal0408** [comment]
@arlakshm @judyjoseph @abdosi could you please help review when get a chance? Thanks!

**Review by arlakshm — APPROVED** [review]

**Comment by deepak-singhal0408** [comment]
Thanks @arlakshm.
@rlhui, could you please help merge this PR? Thanks!

**Comment by gechiang** [comment]
@deepak-singhal0408 Cherry-pick conflict for MSFT repo 202205 branch.  Please raise PR directly to MSFT repo and mention this PRin the new PR.

**Comment by prsunny** [comment]
@deepak-singhal0408 , seems like swss PR checks are failing with this docker change. Could you confirm all swss tests were tested with this docker vs build?

**Comment by deepak-singhal0408** [comment]
> @deepak-singhal0408 , seems like swss PR checks are failing with this docker change. Could you confirm all swss tests were tested with this docker vs build?

@prsunny , this PR is merged in sonic-buildimage on April 19th. And there is a PR merge in sonic-swss on April 22nd(https://github.com/sonic-net/sonic-swss/pull/3118). The PR checker for April 22nd change would have already taken my changes.. right? May I know why you think this PR would have caused issue?

**Comment by ishidawataru** [comment]
@deepak-singhal0408 Where can I find `vs_chassis_metadata.json`? I built the multi-asic vs image from this repo but there is no `vs_chassis_metadata.json` under `/etc/sonic`.

**Comment by yutongzhang-microsoft** [comment]
Hi, @deepak-singhal0408 , I got the same error
```
Jun 21 03:01:42 vlab-01 determine-reboot-cause[15110]:   File "/usr/lib/python3/dist-packages/sonic_platform/chassis.py", line 26, in __init__
Jun 21 03:01:42 vlab-01 determine-reboot-cause[15110]:     self.metadata = self._read_metadata()
Jun 21 03:01:42 vlab-01 determine-reboot-cause[15110]:                     ^^^^^^^^^^^^^^^^^^^^^
Jun 21 03:01:42 vlab-01 determine-reboot-cause[15110]:   File "/usr/lib/python3/dist-packages/sonic_platform/chassis.py", line 34, in _read_metadata
Jun 21 03:01:42 vlab-01 determine-reboot-cause[15110]:     raise FileNotFoundError("Metadata file {} not found".format(self.metadata_file))
Jun 21 03:01:42 vlab-01 determine-reboot-cause[15110]: FileNotFoundError: Metadata file /etc/sonic/vs_chassis_metadata.json not found
```

**Comment by deepak-singhal0408** [comment]
> Hi, @deepak-singhal0408 , I got the same error
> 
> ```
> Jun 21 03:01:42 vlab-01 determine-reboot-cause[15110]:   File "/usr/lib/python3/dist-packages/sonic_platform/chassis.py", line 26, in __init__
> Jun 21 03:01:42 vlab-01 determine-reboot-cause[15110]:     self.metadata = self._read_metadata()
> Jun 21 03:01:42 vlab-01 determine-reboot-cause[15110]:                     ^^^^^^^^^^^^^^^^^^^^^
> Jun 21 03:01:42 vlab-01 determine-reboot-cause[15110]:   File "/usr/lib/python3/dist-packages/sonic_platform/chassis.py", line 34, in _read_metadata
> Jun 21 03:01:42 vlab-01 determine-reboot-cause[15110]:     raise FileNotFoundError("Metadata file {} not found".format(self.metadata_file))
> Jun 21 03:01:42 vlab-01 determine-reboot-cause[15110]: FileNotFoundError: Metadata file /etc/sonic/vs_chassis_metadata.json not found
> ```

Hi @yutongzhang-microsoft , please give a try with this Fix. https://github.com/sonic-net/sonic-host-services/pull/133
JFYI, this file is not expected to be present on pizza box vs platforms.
Even on chassis based vs image, this is not mandatory to be present..

**Comment by ishidawataru** [comment]
> Even on chassis based vs image, this is not mandatory to be present..

How about stopping raising an exception here instead of catching `FileNotFoundError` everywhere the exception can be ignored if the metadata file is not mandatory?

https://github.com/sonic-net/sonic-buildimage/pull/18512/files#diff-1837abe4216c07096db3b47b74a8ce47b0097622773caf92e9dc9f3c42636d06R34

**Comment by yutongzhang-microsoft** [comment]
> > Even on chassis based vs image, this is not mandatory to be present..
> 
> How about stopping raising an exception instead of catching `FileNotFoundError` everywhere the exception can be ignored if the metadata file is not mandatory?
> 
> https://github.com/sonic-net/sonic-buildimage/pull/18512/files#diff-1837abe4216c07096db3b47b74a8ce47b0097622773caf92e9dc9f3c42636d06R34

I agree with you, some other commands also failed because of this error. 
@deepak-singhal0408 Can you fix as suggested?

**Comment by chahibi** [comment]
@deepak-singhal0408 I am still seeing the same error on the latest image from today. Do you have any update?

**Pull request merged**